### PR TITLE
Add Proxy Attestation example

### PR DESCRIPTION
**Pull request by ipetr0v**
This change adds Proxy Attestation example.
It consists of:
- `experimental/proxy_attestation`: CA that can create signed certificates based on certificate signing requests
- `examples/proxy_attestation/client`: Client that can request the CA root certificate and use it to connect to an Oak application
- `examples/proxy_attestation/module`: Simple example Oak application

Certificate signing requests are created by `oak_loader`.

Fixes https://github.com/project-oak/oak/issues/1860
Ref https://github.com/project-oak/oak/issues/1374

### CODE REVIEW

**Comment by ipetr0v** [comment]
So looks like OpenSSL-Rust doesn't support WASM yet: https://github.com/rustwasm/wasm-bindgen/issues/1529

**Comment by tiziano88** [comment]
Why are you trying to use SSL from Wasm?

**Comment by ipetr0v** [comment]
I needed a crate that could create x509 certificates based on a certificate request.

**Comment by tiziano88** [comment]
But why in Wasm?

**Comment by ipetr0v** [comment]
Because we need to create it in Oak. I.e. backend server sends Oak a certificate request, and Oak creates it using its root key.

**Comment by tiziano88** [comment]
That's not what I was expecting. The CA would be a separate stand-alone binary that has nothing to do with Oak. Its job is to verify remote attestation quotes, and sign certs based on that. The impl of the gRPC server pseudo node would obtain such signature from the CA enclave, and then start listening over gRPC using that cert. No Wasm is involved, and definitely no other Oak-specific functionality.

What is your plan instead?

**Comment by ipetr0v** [comment]
I thought the original idea was to make clients trust that the backend was attested and correct certificates were signed. And in order to enforce that, the CA and the attestor should be in Oak (or at least as a separate app in TEE), so they themselves could be remotely attestable (by clients).

And I tried to make it all in Wasm just to have a Proxy Attestation Service as an Oak application, and not making it a part of Runtime.

**Comment by tiziano88** [comment]
> I thought the original idea was to make clients trust that the backend was attested and correct certificates were signed. And in order to enforce that, the CA and the attestor should be in Oak (or at least as a separate app in TEE), so they themselves could be remotely attestable (by clients).

How would this make them remotely attestable? In fact I think this makes it impossible for a client to make sure that a particular application is actually running in Oak. And additionally, how would the application remotely attest to the CA? Even if it works, the layering of Oak runtime + Oak application seems completely unnecessary and overengineered, and makes the TCB of the CA massive. I thought we agreed that the solution should be minimal and independent of Oak.

**Comment by ipetr0v** [comment]
So neither the CA nor a backend app should use Oak, right?

Also, in the case when a backup application uses Oak, it would still need to create a certificate signing request (using something different from `openssl-rust`).

**Comment by tiziano88** [comment]
> So neither the CA nor a backend app should use Oak, right?

Correct. The goal is to have a proxy for remote attestation, no more, no less.

> Also, in the case when a backup application uses Oak, it would still need to create a certificate signing request (using something different from `openssl-rust`).

I am not sure what you mean by "backup application", could you clarify?

**Comment by ipetr0v** [comment]
> I am not sure what you mean by "backup application", could you clarify?

I have written "backup" instead of "backend" :^)
But basically, it's the application that wants to be attested by the proxy.

**Comment by tiziano88** [comment]
The Oak loader will generate a cert / key, attest itself to the proxy, get its certificate signed by the proxy (which verifies the remote attestation claim with Intel / AMD, for instance), then use that cert / key to terminate TLS for the gRPC connection of a gRPC server pseudo-node.

This way a client connecting to the Oak loader (running any Oak application) would know that it is talking to a VM that has been remotely attested by the CA enclave. There is still the question of what exactly does the CA enclave do when signing the worker enclave cert, e.g. does it need to put the remote attestation quote in a special field, or does it store it locally, or something else?

**Comment by ipetr0v** [comment]
I don't think the quote should be stored by the Proxy, because after the attestation this quote is only needed by the client - to check that it's talking to a specific application that they intended to. So this quote should be in the TLS certificate used by the backend application.

Also, we probably don't even need a quote, because clients do not operate with backend TEE credentials (such as EPID remote attestation keys and etc.) of the backend application. In our case `MRENCLAVE` or `Image Hash` will be enough.

**Comment by tiziano88** [comment]
> I don't think the quote should be stored by the Proxy, because after the attestation this quote is only needed by the client - to check that it's talking to a specific application that they intended to. So this quote should be in the TLS certificate used by the backend application.

Yes, but now we need to find a field in the TLS cert to store it. BTW, have you looked at how Intel does this in SGX? I think we talked about the fact that we should try and be compatible with what they do, if at all possible.

> Also, we probably don't even need a quote, because clients do not operate with backend TEE credentials (such as EPID remote attestation keys and etc.) of the backend application. In our case `MRENCLAVE` or `Image Hash` will be enough.

Why limit to a single field? Also I don't think it will be enough. For instance, a client may want to know the SGX version number, or whether the enclave is in debug mode, or anything else. We should just leave the quote as it is, and replace its signature with that of the CA enclave, which is also conceptually a simple operation. CA enclave gets a quote, checks the remote attestation signature, then throws it away and replaces it with its own signature.

**Comment by conradgrobler** [comment]
> BTW, have you looked at how Intel does this in SGX?

FYI, I think this has quite a bit of detail on the topic: https://arxiv.org/pdf/1801.05863.pdf

File: oak_loader/src/attestation.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+/// .
+fn create_certificate_request(key_pair: &PKey<Private>) -> anyhow::Result<Vec<u8>> {
+    let mut builder = X509ReqBuilder::new()?;
+    builder.set_pubkey(&key_pair)?;
+
+    let mut cert_name = X509NameBuilder::new()?;
+    // CN refers to Common Name.
+    // https://tools.ietf.org/html/rfc5280
+    cert_name.append_entry_by_text("CN", &get_tee_measurements())?;
```
Comment (tiziano88) [inline oak_loader/src/attestation.rs:80]: What makes sure that this is putting a real measurement in this field? AFAICT this is completely unrelated to the actual attestation with the proxy, so why would anyone trust it?

Comment (ipetr0v) [reply oak_loader/src/attestation.rs:80]: The proxy will attest the backend and only sign a certificate that has a correct measurement in it.
But proxy probably should be the one who puts these measurements in the cert (in the end it will be identical cert).

Comment (tiziano88) [reply oak_loader/src/attestation.rs:82]: Is it not possible for the CA enclave to add those fields and sign them in its part of the cert, instead of double checking them from the self-signed cert from the worker enclave?

File: examples/proxy_attestation/module/rust/build.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright 2020 The Project Oak Authors
+// Copyright 2021 The Project Oak Authors
```
Comment (daviddrysdale) [inline examples/proxy_attestation/module/rust/build.rs:2]: Random drive-by: leave copyright date as [date of creation](https://opensource.google/docs/copyright/#the-year), not date of last modification.

Comment (ipetr0v) [reply examples/proxy_attestation/module/rust/build.rs:2]: It's a newly created file

Comment (daviddrysdale) [reply examples/proxy_attestation/module/rust/build.rs:2]: Ah, my bad, I made assumptions because it was showing as a diff.

File: experimental/proxy_attestation/src/certificate.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        builder.append_extension(auth_key_identifier)?;
+
+        let subject_alt_name = SubjectAlternativeName::new()
+            .dns("project-oak.local")
+            .dns("localhost")
+            .ip("127.0.0.1")
+            .ip("::1")
+            // `10.0.2.2` is routed to the host machine by the Android emulator.
+            .ip("10.0.2.2")
+            .build(&builder.x509v3_context(Some(&self.root_certificate), None))?;
```
Comment (tiziano88) [inline experimental/proxy_attestation/src/certificate.rs:185]: Do these need to be the actual host names / IPs of the application?

File: experimental/proxy_attestation/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+};
+use openssl::x509::X509Req;
+use structopt::StructOpt;
+use tonic::{
+    transport::{Identity, Server, ServerTlsConfig},
+    Request, Response, Status,
+};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Aggregator Backend")]
```
Comment (tiziano88) [inline experimental/proxy_attestation/src/main.rs:61]: Not sure whether these comments are accurate.

File: experimental/proxy_attestation/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+//!
+//! Proxy Attestation Service works as a Certificate Authority for TEE applications. It attests
+//! applications and creates signed certificates for TEE applications, which they use to create TLS
+//! connections with clients.
+//!
+//! The work cycle of the proxy is represented in 2 stages.
+//! First stage corresponds to the backend server attestation:
+//! - TEE application connects to the proxy
+//! - Proxy attests the application using a corresponding remote attestation protocol
+//! - Application sends the proxy a certificate signing request that contains TEE measurements
```
Comment (tiziano88) [inline experimental/proxy_attestation/src/main.rs:27]: I don't think the client needs to send any measurement, besides what's already in the remote attestation step. Otherwise now we have the real measurement, and also the one in this request, which seems unnecessary and confusing (and I'm sure will lead to bugs).

File: experimental/proxy_attestation/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+//! - Proxy sings it using its private key and sends the signature back to the application
```
Comment (tiziano88) [inline experimental/proxy_attestation/src/main.rs:28]:
```suggestion
//! - Proxy signs the application certificate using its private key and sends the signed certificate back to the application
```

File: experimental/proxy_attestation/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+//!
+//! Second stage corresponds to the client connection:
+//! - Client remotely attests the proxy using a corresponding attestation protocol
+//! - Proxy sends the client its root certificate
+//!     - Client trusts that this is a correct certificate, since it was sent via a secure channel
+//!       created during the attestation process
+//! - Client connects to the application using TLS, and the application uses a TLS certificate
+//!   previously signed by the proxy
+//! - Client checks that the certificate was signed by the root certificate and establishes a secure
+//!   connection
```
Comment (tiziano88) [inline experimental/proxy_attestation/src/main.rs:38]: At some point here I guess the client actually checks the remote attestation details on the certificate?

File: experimental/proxy_attestation/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+            Ok(certificate) => match certificate.to_pem() {
+                Ok(serialized_certificate) => {
+                    info!("Sending signed certificate");
+                    let mut res = GetSignedCertificateResponse::default();
+                    res.certificate = serialized_certificate;
+                    Ok(Response::new(res))
+                }
+                Err(error) => Err(Status::internal(format!(
+                    "Certificate serialization error: {}",
+                    error
```
Comment (tiziano88) [inline experimental/proxy_attestation/src/main.rs:112]: Please don't just send unsanitized error messages to clients like this, they may contain all sort of internal data. Also, it seems that the server is not even logging these errors. I think exactly the opposite should happen: log on the server, and send a generic message back to the client. Otherwise it is basically impossible to debug.

File: oak_loader/src/attestation.rs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+/// Creates an X.509 certificate signing request.
+/// https://tools.ietf.org/html/rfc5280
+fn create_certificate_request(key_pair: &PKey<Private>) -> anyhow::Result<Vec<u8>> {
+    let mut builder = X509ReqBuilder::new()?;
+    builder.set_pubkey(&key_pair)?;
+
+    let mut name = X509NameBuilder::new()?;
+    name.append_entry_by_text("O", "Oak")?;
+    name.append_entry_by_text("CN", "Example Application")?;
```
Comment (tiziano88) [inline oak_loader/src/attestation.rs:77]: Are these final values or placeholders?

File: experimental/proxy_attestation/src/main.rs (hunk 5 above)
Comment (conradgrobler) [reply experimental/proxy_attestation/src/main.rs:27]: +1. The measurement should be the same as in the initial attestation of the application to the proxy.

File: experimental/proxy_attestation/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+use structopt::StructOpt;
+use tonic::{
+    transport::{Identity, Server, ServerTlsConfig},
+    Request, Response, Status,
+};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Aggregator Backend")]
+pub struct Opt {
+    #[structopt(long, help = "Private RSA key file used by gRPC server.")]
```
Comment (conradgrobler) [inline experimental/proxy_attestation/src/main.rs:63]: What format is this file in?

File: experimental/proxy_attestation/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+//!
+//! The work cycle of the proxy is represented in 2 stages.
+//! First stage corresponds to the backend server attestation:
+//! - TEE application connects to the proxy
+//! - Proxy attests the application using a corresponding remote attestation protocol
+//! - Application sends the proxy a certificate signing request that contains TEE measurements
+//! - Proxy sings it using its private key and sends the signature back to the application
+//!
+//! Second stage corresponds to the client connection:
+//! - Client remotely attests the proxy using a corresponding attestation protocol
```
Comment (conradgrobler) [inline experimental/proxy_attestation/src/main.rs:31]: What is the plan for this attestation protocol? Modified TLS? EKEP? Something else?

File: experimental/proxy_attestation/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+//! Proxy Attestation Service implementation.
+//!
+//! Proxy Attestation Service works as a Certificate Authority for TEE applications. It attests
+//! applications and creates signed certificates for TEE applications, which they use to create TLS
+//! connections with clients.
+//!
+//! The work cycle of the proxy is represented in 2 stages.
+//! First stage corresponds to the backend server attestation:
+//! - TEE application connects to the proxy
+//! - Proxy attests the application using a corresponding remote attestation protocol
```
Comment (conradgrobler) [inline experimental/proxy_attestation/src/main.rs:26]: Is the plan to use EKEP here? I don't think a modified TLS works unless it does modified mutual attesation, since the application is initiating the connection.

File: experimental/proxy_attestation/src/main.rs (hunk 4 above)
Comment (ipetr0v) [reply experimental/proxy_attestation/src/main.rs]: Done

File: experimental/proxy_attestation/src/main.rs (hunk 10 above)
Comment (ipetr0v) [reply experimental/proxy_attestation/src/main.rs]: It's PEM encoded.
Added a comment.

File: experimental/proxy_attestation/src/main.rs (hunk 8 above)
Comment (ipetr0v) [reply experimental/proxy_attestation/src/main.rs]: Done

File: oak_loader/Cargo.toml [hunk 13]
```diff
@@ -21,16 +21,22 @@ default = ["oak_debug"]
 anyhow = "*"
 env_logger = "*"
 hex = "*"
+http = "*"
 log = "*"
 oak_abi = { path = "../oak_abi" }
 oak_runtime = { path = "../oak_runtime", default-features = false }
+oak_proxy_attestation = { path = "../experimental/proxy_attestation" }
```
Comment (conradgrobler) [inline oak_loader/Cargo.toml:28]: I think it would be a good idea to add an "oak_experimental" feature that is not enabled by default so that the oak server only depends on experimental crates (and their dependencies) if it is explicitly requested to be an experimental build.

Comment (ipetr0v) [reply oak_loader/Cargo.toml:28]: Added an `oak_attestation` feature, and it's running as part of the `--server-variant=experimental` which is now default in CI.

**Review by conradgrobler — APPROVED** [review]
